Mono.Nat/AsyncExtensions.cs

```diff
@@ -57,12 +57,31 @@ public static async void FireAndForget (this Task task)
         public static void WaitAndForget (this Task task)
         {
             try {
-                task.GetAwaiter ().GetResult ();
+                task.ConfigureAwait (false).GetAwaiter().GetResult();
             } catch (OperationCanceledException) {
                 // If we cancel the task then we don't need to log anything.
             } catch (Exception ex) {
                 Log.ErrorFormatted ("Unhandled exception: {0}{1}", Environment.NewLine, ex);
             }
         }
+        /// <summary>
+        /// Adds cancellation functionality to a task that does not accept a CancellationToken otherwise.
+        /// https://stackoverflow.com/questions/19404199/how-to-to-make-udpclient-receiveasync-cancelable#:~:text=There's%20no%20built%2Din%20way,Delay()%20to%20implement%20timeouts).
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="task"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public static async Task<T> WithCancellation<T> (this Task<T> task, CancellationToken cancellationToken)
```

Comment:
This method is a poor way to implement cancellation, as it doesn't actually cancel the underlying task.

In this particular scenario, we cannot have faux cancellation like this as it leaves the library in a state where multiple concurrent `SendAsync` calls could be made on the same object (an error), or the same for ReceiveAsync.

Reply:
I dunno; if it leads to an exception being thrown rather than the app just hanging like it was, then I'd call it a win.  But then, like I said, I'm very inexperienced with async programming in C# and have little patience with it :)

Reply:
I'd definitely agree with this! An exception is far better than a deadlock.

I went and solved this a slightly different way (removing all the synchronous blocking code) so we should be problem free now!